File: day12/sort_algorithm/08_countingSort.java
class Solution {
        
    public int[] sortArray(int[] nums) {
        if (nums.length > 1) {

            int nmax = nums[0];
            int nmin = nums[0];
            for (int num: nums) {
                if (num > nmax) nmax = num;
                if (num < nmin) nmin = num;
            }
            int n = nmax - nmin + 1;

            var arr = new int[n];
            for (int num: nums) {
                arr[num - nmin]++;
            }

            int pivot = 0;
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < arr[i]; j++) {
                    nums[pivot] = nmin + i;
                    pivot++;
                }
            }
        }
        return nums;
    }
}

/*
class Solution {
    public int[] sortArray(int[] nums) {
        if (nums.length > 1) {
            int max = nums[0], min = nums[0];
            for (int n: nums) {
                if (max < n) max = n;
                if (min > n) min = n;
            }
            int[] count = new int[max - min + 1];
            for (int n: nums) {
                count[n - min]++;
            }
            for (int i = 1; i < count.length; i++) {
                count[i] += count[i - 1];
            }

            int[] res = new int[nums.length];
            for (int i = nums.length - 1; i > -1; i--) {
                res[--count[nums[i] - min]] = nums[i];
            }
            nums = res;
        }
        return nums;
    }
}
*/

File: day12/sort_algorithm/08_counting_sort.py
class Solution:
    def counting_sort(self, nums):
        if nums:
            nmax = max(nums) + 1
            nmin = min(nums)

            counting_arr = [0] * (nmax - nmin)
            for num in nums:
                counting_arr[num - nmin] += 1
            
            pivot = 0
            for i in range(len(counting_arr)):
                for j in range(counting_arr[i]):
                    nums[pivot] = nmin + i
                    pivot += 1
        return nums

    def sortArray(self, nums: List[int]) -> List[int]:
        return self.counting_sort(nums)
